tools/terraform-bundle/package.go

```diff
@@ -117,86 +68,95 @@ func (c *PackageCommand) Run(args []string) int {
-	for k, _ := range localPlugins {
-		c.ui.Info(fmt.Sprintf("plugin: %s (%s)", k.Name, k.Version))
+	// get the list of required providers from the config
+	reqs := make(map[addrs.Provider][]string)
+	for name, provider := range config.Providers {
+		var fqn addrs.Provider
+		var diags tfdiags.Diagnostics
+		if provider.Source != "" {
+			fqn, diags = addrs.ParseProviderSourceString(provider.Source)
+			if diags.HasErrors() {
+				c.ui.Error(fmt.Sprintf("Invalid provider source string: %s", provider.Source))
+				return 1
+			}
+		} else {
+			fqn = addrs.NewDefaultProvider(name)
+		}
+		reqs[fqn] = provider.Versions
 	}
-	installer := &discovery.ProviderInstaller{
-		Dir: workDir,
-
-		// FIXME: This is incorrect because it uses the protocol version of
-		// this tool, rather than of the Terraform binary we just downloaded.
-		// But we can't get this information from a Terraform binary, so
-		// we'll just ignore this for now and use the same plugin installer
-		// protocol version for terraform-bundle as the terraform shipped
-		// with this release.
-		//
-		// NOTE: To target older versions of terraform, use the terraform-bundle
-		// from the same tag.
-		PluginProtocolVersion: discovery.PluginInstallProtocolVersion,
 
+	// set up the provider installer
+	platform := getproviders.Platform{
 		OS:   osName,
 		Arch: archName,
-		Ui:   c.ui,
 	}
-
-	for name, constraintStrs := range config.Providers {
-		for _, constraintStr := range constraintStrs {
-			c.ui.Output(fmt.Sprintf("- Resolving %q provider (%s)...",
-				name, constraintStr))
-			foundPlugins := discovery.PluginMetaSet{}
-			constraint := constraintStr.MustParse()
-			for plugin, _ := range localPlugins {
-				if plugin.Name == name && constraint.Allows(plugin.Version.MustParse()) {
-					foundPlugins.Add(plugin)
-				}
-			}
-
-			if len(foundPlugins) > 0 {
-				plugin := foundPlugins.Newest()
-				CopyFile(plugin.Path, workDir+"/terraform-provider-"+plugin.Name+"_v"+plugin.Version.MustParse().String()) //put into temp dir
-			} else { //attempt to get from the public registry if not found locally
-				c.ui.Output(fmt.Sprintf("- Checking for provider plugin on %s...",
-					releaseHost))
-				_, _, err := installer.Get(addrs.NewLegacyProvider(name), constraint)
-				if err != nil {
-					c.ui.Error(fmt.Sprintf("- Failed to resolve %s provider %s: %s", name, constraint, err))
-					return 1
+	installdir := providercache.NewDirWithPlatform(filepath.Join(workDir, "plugins"), platform)
+
+	services := disco.New()
+	services.SetUserAgent(httpclient.TerraformUserAgent(version.String()))
+	var sources []getproviders.MultiSourceSelector
+
+	// Find any local providers first so we can exclude these from the registry
+	// install. We'll just silently ignore any errors and assume it would fail
+	// real installation later too.
+	foundLocally := map[addrs.Provider]struct{}{}
+
+	if absPluginDir, err := filepath.Abs(pluginDir); err == nil {
+		if _, err := os.Stat(absPluginDir); err == nil {
```

Comment:
Some logging around here might be helpful. Examples of things that would've helped me find my own mistake today

- `No ./.plugins directory found, skipping local provider discovery.`
- `Found ./.plugins directory, discovering local providers.`
- `Local provider found: "example.com/myorg/customplugin"`
- `No local providers found.`

Reply:
love it, added all of them

Reply:
I hope you like that commit message 😝